### PR TITLE
Finish SAD reorganization and make it the default guess

### PR DESCRIPTION
**Pull request by susilehtola**
## Description
This PR finishes the reorganization of the SAD code. Renames ```SAD_FRAC_SR_OCC``` to ```SAD_SPIN_AVERAGE```, which is also valid without fractional occupations. Fixes the guess in the presence of ghost atoms.

I also turn fractional occupations and spin-averaging on by default, because
* fractional occupations are well-known to give better results than integer ones
* differences between spin-averaged SAD and atomic ground state SAD appear to be small when fractional occupations are used
* the spin-averaged code not limited to Z <= Rn

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Turn fractional averaging in SAD on by default
- [x] Turn spin-averaging in SAD on by default
- [x] SAD works in the presence of ghost atoms
- [x] SAD works in the presence of ECPs
- [x] Make SAD the default guess 

## Questions
- [ ] Question1

## Checklist
- [ ] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by loriab** [comment]
Quick question – have you run a full & clean test suite with frac occ on by default? I ask because I tried turning it on unilaterally once before and it fixed my troublesome case but broke some tests. I expect your recent PRs have fixed things, but it's good to check the baseline. Thanks!

**Comment by susilehtola** [comment]
@loriab full tests are running, although I have the default libint which is handicapped. Your failures were most likely caused by #1451: since the occupation numbers weren't included, e.g. the Sc atom would have been calculated in the charge state -15(!).

File: tests/scf-bs/input.dat [hunk 1]
```diff
@@ -11,6 +11,7 @@ H 1 3.0
 symmetry c1
 }
 set reference uhf
+set guess gwh
```
Comment (susilehtola) [inline tests/scf-bs/input.dat]: So this is a UHF calculation on H2 close to dissociation.

I looked at the code, and the symmetry breaking should also work with SAD, since it is done when the orbitals are formed for the first time.

What must be happening here is that the HOMO and LUMO are of the same symmetry in the SAD guess, so mixing these orbitals does not give you the wanted broken symmetry.

**Comment by susilehtola** [comment]
Other than seeing python test failures caused by psi4 being compiled with python3 not python2 assumed by the tests, nothing big comes out.

dft-grad-meta is failing, but that might be due to me using a different version of libxc. Some tests are also failing for me due to the default libint parameters.

If you have a setup that can run all tests, you might want to double check.

**Comment by loriab** [comment]
> Other than seeing python test failures caused by psi4 being compiled with python3 not python2 assumed by the tests, nothing big comes out.

I hope these are tests in `tests/python/` and `tests/json/` you're seeing this with (those are psiapi)? All psithon `ctest` tests should be using the python psi was built with, not the `which python` python. 

Sure, I'll run a local full tests on this PR this weekend.

**Comment by susilehtola** [comment]
The benefits of the SAD guess are pretty humongous over the current default. E.g. 
```
molecule {
0 2
V 0.000000 0.000000 0.000000 
N -0.009654 1.561722 1.062138 
N 0.009654 -1.561722 1.062138 
N -1.561722 -0.009654 -1.062138 
N 1.561722 0.009654 -1.062138 
C -1.219978 2.113945 1.655979 
C 1.219978 -2.113945 1.655979 
C -2.113945 -1.219978 -1.655979 
C 2.113945 1.219978 -1.655979 
C 1.188278 2.080402 1.709075 
C -1.188278 -2.080402 1.709075 
C -2.080402 1.188278 -1.709075 
C 2.080402 -1.188278 -1.709075 
H -2.112996 1.691813 1.177356 
H 2.112996 -1.691813 1.177356 
H -1.691813 -2.112996 -1.177356 
H 1.691813 2.112996 -1.177356 
H 2.089441 1.627028 1.275983 
H -2.089441 -1.627028 1.275983 
H -1.627028 2.089441 -1.275983 
H 1.627028 -2.089441 -1.275983 
H -1.254235 3.214541 1.536136 
H 1.254235 -3.214541 1.536136 
H -1.905317 -1.281383 -2.742242 
H 1.905317 1.281383 -2.742242 
H 1.192045 1.877111 2.798018 
H -1.192045 -1.877111 2.798018 
H -3.178469 1.263762 -1.584921 
H 3.178469 -1.263762 -1.584921 
H -1.281383 1.905317 2.742242 
H 1.281383 -1.905317 2.742242 
H -3.214541 -1.254235 -1.536136 
H 3.214541 1.254235 -1.536136 
H 1.263762 3.178469 1.584921 
H -1.263762 -3.178469 1.584921 
H -1.877111 1.192045 -2.798018 
H 1.877111 -1.192045 -2.798018
}

set basis pcseg-1
set reference uhf
set scf_type direct
set sad_print 2
set guess sad
set sad_spin_average false
set df_scf_guess false
energy('scf')
```

| Guess | Num iter | Energy |
| -- | -- | -- |
| fractional SAD, ground spin state | 13 | -1477.26833028686269 |
| fractional SAD, spin-averaged occupations | 14 | -1477.26833044972045 |
| gwh | 62 | -1477.12773301260950 |
| core | >100 | |

so not only is the number of iterations reduced by almost a factor of 5, it also yields a lower solution.

**Comment by susilehtola** [comment]
@loriab yup, failures are for ```python-*/```, ```json-*```, as well as ```psi4numpy-*```

**Comment by dgasmith** [comment]
You could be landing on a different minima there? HCN+ was chosen mostly because it was small, but it might be too unstable for a test.

**Comment by susilehtola** [comment]
@dgasmith that's what I thought. I bet the problem is just that since HCN is a linear molecule, HCN+ probably has a hole in one of the pi orbitals, and then it's a game of change which one of the two you converge upon since they're fully degenerate. I changed the tests to SSH which should behave better.

**Comment by susilehtola** [comment]
All tests work at my end, except the python ones that fail due to the wrong interpreter like stated above.

**Comment by susilehtola** [comment]
Looks like it's failing on travis in
```stage/lib/psi4/tests/test_vibanalysis.py::test_hessian_vs_cfour[methane-H_analytic] FAILED [ 93%]```

so the analytic Hessian for CH4 is failing, but the finite difference calculations based on forces or energies are both okay... I don't really see what could be going wrong here: if it's doing the neutral molecule, this PR hasn't changed anything!

**Comment by loriab** [comment]
Pretty sure it’s just an eigenvector ordering amongst degen freq instability that appears rarely. Wanted to test locally before restart.

**Comment by susilehtola** [comment]
Ohhh that’s right; if it’s checking eigenvectors then you can get funny results with (near) degeneracies.. do you have any suggestions on how to fix it?

Susi Lehtola
Sent from my phone so excuse my brevity.


On 5 Jan 2019, at 21.08, Lori A. Burns <notifications@github.com<mailto:notifications@github.com>> wrote:


Pretty sure it’s just an eigenvector ordering amongst degen freq instability that appears rarely. Wanted to test locally before restart.

**Comment by loriab** [comment]
Rest of a full test suite is clean, but the methane analytic freq failing for me locally, too.

Judging from https://travis-ci.org/psi4/psi4/jobs/476480673#L2805 and similar, the eigenvector alignment is actually working properly (the diffs would show values like .3 if not) and it's ~0.02 differences in the eigenvectors themselves that are the trouble.

```
Degenerate modes reordered: 0-->2, 2-->0, 3-->4, 4-->3, 6-->8, 8-->6
Degenerate modes reordered: 1-->2, 2-->1
q : False
	exp: 
 [[-0.0000 -0.3972 -0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000 -0.3061]
 [ 0.0000 -0.0000 -0.3972  0.0000 -0.0000 -0.0000  0.0000 -0.3061 -0.0000]
 [ 0.3972 -0.0000 -0.0000 -0.0000  0.0000 -0.0000  0.3061 -0.0000 -0.0000]
 [ 0.0000  0.5585 -0.0000  0.0000  0.5000 -0.0000 -0.0000 -0.0000 -0.0160]
 [-0.3052 -0.0000  0.1268  0.2887  0.0000  0.4082  0.3960  0.5441 -0.0000]
 [-0.3426 -0.0000  0.3052  0.4082  0.0000 -0.2887 -0.2641 -0.3960 -0.0000]
 [ 0.0000  0.5585 -0.0000  0.0000 -0.5000 -0.0000 -0.0000  0.0000 -0.0160]
 [ 0.3052 -0.0000  0.1268 -0.2887  0.0000 -0.4082 -0.3960  0.5441 -0.0000]
 [-0.3426 -0.0000 -0.3052  0.4082  0.0000 -0.2887 -0.2641  0.3960 -0.0000]
 [-0.3052  0.1268  0.0000 -0.2887  0.0000 -0.4082  0.3960 -0.0000  0.5441]
 [ 0.0000 -0.0000  0.5585  0.0000 -0.5000  0.0000  0.0000 -0.0160 -0.0000]
 [-0.3426  0.3052 -0.0000 -0.4082  0.0000  0.2887 -0.2641 -0.0000 -0.3960]
 [ 0.3052  0.1268 -0.0000  0.2887  0.0000  0.4082 -0.3960  0.0000  0.5441]
 [ 0.0000 -0.0000  0.5585  0.0000  0.5000  0.0000  0.0000 -0.0160 -0.0000]
 [-0.3426 -0.3052 -0.0000 -0.4082  0.0000  0.2887 -0.2641  0.0000  0.3960]]
	obs: 
 [[ 0.0002 -0.3972 -0.0007  0.0000 -0.0000 -0.0000 -0.0008 -0.0128 -0.3058]
 [ 0.0017  0.0007 -0.3972 -0.0000 -0.0000  0.0000 -0.0000 -0.3058  0.0128]
 [ 0.3972  0.0002  0.0017 -0.0000  0.0000 -0.0000  0.3061 -0.0000 -0.0008]
 [-0.0003  0.5585  0.0010 -0.0000  0.5000 -0.0000 -0.0000 -0.0007 -0.0159]
 [-0.3058 -0.0004  0.1255  0.2887  0.0000  0.4082  0.3960  0.5436 -0.0238]
 [-0.3439 -0.0007  0.3038  0.4082  0.0000 -0.2887 -0.2641 -0.3956  0.0172]
 [-0.0003  0.5585  0.0010  0.0000 -0.5000 -0.0000 -0.0000 -0.0007 -0.0159]
 [ 0.3047 -0.0001  0.1281 -0.2887 -0.0000 -0.4082 -0.3960  0.5437 -0.0218]
 [-0.3413  0.0004 -0.3067  0.4082  0.0000 -0.2887 -0.2641  0.3957 -0.0159]
 [-0.3053  0.1267 -0.0011 -0.2887 -0.0000 -0.4082  0.3974  0.0227  0.5426]
 [-0.0024 -0.0010  0.5585  0.0000 -0.5000  0.0000 -0.0000 -0.0159  0.0007]
 [-0.3428  0.3051 -0.0009 -0.4082 -0.0000  0.2887 -0.2651 -0.0166 -0.3950]
 [ 0.3052  0.1270  0.0015  0.2887  0.0000  0.4082 -0.3947  0.0228  0.5446]
 [-0.0024 -0.0010  0.5585 -0.0000  0.5000  0.0000 -0.0000 -0.0159  0.0007]
 [-0.3425 -0.3054 -0.0020 -0.4082 -0.0000  0.2887 -0.2631  0.0166  0.3963]]
	dif: 
 [[ 0.0002  0.0000 -0.0007  0.0000 -0.0000 -0.0000 -0.0008 -0.0128  0.0003]
 [ 0.0017  0.0007  0.0000 -0.0000 -0.0000  0.0000 -0.0000  0.0003  0.0128]
 [-0.0000  0.0002  0.0017 -0.0000  0.0000 -0.0000 -0.0000 -0.0000 -0.0008]
 [-0.0003 -0.0000  0.0010 -0.0000 -0.0000 -0.0000 -0.0000 -0.0007  0.0000]
 [-0.0005 -0.0004 -0.0013  0.0000  0.0000 -0.0000  0.0000 -0.0005 -0.0238]
 [-0.0013 -0.0007 -0.0015 -0.0000  0.0000 -0.0000 -0.0000  0.0004  0.0172]
 [-0.0003 -0.0000  0.0010  0.0000  0.0000 -0.0000 -0.0000 -0.0007  0.0000]
 [-0.0005 -0.0001  0.0013 -0.0000 -0.0000  0.0000  0.0000 -0.0004 -0.0218]
 [ 0.0013  0.0004 -0.0015 -0.0000  0.0000 -0.0000  0.0000 -0.0003 -0.0159]
 [-0.0001 -0.0001 -0.0011 -0.0000 -0.0000  0.0000  0.0014  0.0227 -0.0015]
 [-0.0024 -0.0010 -0.0000  0.0000  0.0000  0.0000 -0.0000  0.0000  0.0007]
 [-0.0001 -0.0002 -0.0009  0.0000 -0.0000  0.0000 -0.0010 -0.0166  0.0010]
 [-0.0001  0.0001  0.0015  0.0000  0.0000 -0.0000  0.0014  0.0228  0.0005]
 [-0.0024 -0.0010 -0.0000 -0.0000 -0.0000  0.0000 -0.0000  0.0000  0.0007]
 [ 0.0002 -0.0002 -0.0020  0.0000 -0.0000  0.0000  0.0010  0.0166  0.0003]]
```

**Comment by susilehtola** [comment]
I think it's pretty clear that this is not a problem in the code - the calculations run and give the correct result - but in the test itself that has degenerate modes which don't appear to be treated properly by the test code.

Because the vibrational analysis part was also marked as ```xfail``` in the test suite, I found it reasonable to mark the Hessian test as ```xfail```, too. There still remain checks on more complicated molecules, which don't have degenerate frequencies.

As cofirmed by @loriab, all tests pass, so this one is good to go!

**Comment by loriab** [comment]
No, I think the trouble is that the Hessian has changed, not a problem with the test code. Look at the clean `0.3061` elements in the upper right 3x3 of the `obs:` eigenvectors below (c. last week) vs. above (this PR). I'm not sure if the guess change is to blame or something after it, but the problem is not the purely in-test evec regularization I first suspected.

```
omega : True
	exp:  [ 1433.2785  1433.2785  1433.2785  1647.8698  1647.8698  3164.8164
  3285.6166  3285.6166  3285.6166]
	obs:  [1433.27806883+0.j 1433.27806887+0.j 1433.27806888+0.j 1647.86973121+0.j
 1647.86973124+0.j 3164.81841753+0.j 3285.61874563+0.j 3285.61874569+0.j
 3285.61874571+0.j]
	dif:  [-4.31173686e-04+0.j -4.31127530e-04+0.j -4.31121772e-04+0.j
 -6.87897177e-05+0.j -6.87593285e-05+0.j  2.01752747e-03+0.j
  2.14562915e-03+0.j  2.14569447e-03+0.j  2.14571466e-03+0.j]
gamma : False
	exp:  ['T2', 'T2', 'T2', 'E', 'E', 'A1', 'T2', 'T2', 'T2']
	obs:  ['B2' None None 'A2' 'A1' 'A1' 'B1' 'A1' 'B2']
	dif: Different, inspect arrays
IR_intensity : True
	exp:  [ 9.8963  9.8963  9.8963  0.0000  0.0000  0.0000  34.4012  34.4012
  34.4012]
	obs:  [ 9.8965  9.8965  9.8965  0.0000  0.0000  0.0000  34.4016  34.4016
  34.4016]
	dif:  [ 0.0002  0.0002  0.0002  0.0000  0.0000  0.0000  0.0004  0.0004  0.0004]
q : True
	exp: 
 [[-0.0000 -0.3972 -0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000 -0.3061]
 [ 0.0000 -0.0000 -0.3972  0.0000 -0.0000 -0.0000  0.0000 -0.3061 -0.0000]
 [ 0.3972 -0.0000 -0.0000 -0.0000  0.0000 -0.0000  0.3061 -0.0000 -0.0000]
 [ 0.0000  0.5585 -0.0000  0.0000  0.5000 -0.0000 -0.0000 -0.0000 -0.0160]
 [-0.3052 -0.0000  0.1268  0.2887  0.0000  0.4082  0.3960  0.5441 -0.0000]
 [-0.3426 -0.0000  0.3052  0.4082  0.0000 -0.2887 -0.2641 -0.3960 -0.0000]
 [ 0.0000  0.5585 -0.0000  0.0000 -0.5000 -0.0000 -0.0000  0.0000 -0.0160]
 [ 0.3052 -0.0000  0.1268 -0.2887  0.0000 -0.4082 -0.3960  0.5441 -0.0000]
 [-0.3426 -0.0000 -0.3052  0.4082  0.0000 -0.2887 -0.2641  0.3960 -0.0000]
 [-0.3052  0.1268  0.0000 -0.2887  0.0000 -0.4082  0.3960 -0.0000  0.5441]
 [ 0.0000 -0.0000  0.5585  0.0000 -0.5000  0.0000  0.0000 -0.0160 -0.0000]
 [-0.3426  0.3052 -0.0000 -0.4082  0.0000  0.2887 -0.2641 -0.0000 -0.3960]
 [ 0.3052  0.1268 -0.0000  0.2887  0.0000  0.4082 -0.3960  0.0000  0.5441]
 [ 0.0000 -0.0000  0.5585  0.0000  0.5000  0.0000  0.0000 -0.0160 -0.0000]
 [-0.3426 -0.3052 -0.0000 -0.4082  0.0000  0.2887 -0.2641  0.0000  0.3960]]
	obs: 
 [[-0.0007 -0.3972 -0.0000  0.0000  0.0000 -0.0000 -0.0000 -0.0000 -0.3061]
 [-0.0000  0.0000 -0.3972  0.0000 -0.0000 -0.0000 -0.0000 -0.3061  0.0000]
 [ 0.3972 -0.0007 -0.0000 -0.0000 -0.0000 -0.0000  0.3061 -0.0000 -0.0000]
 [ 0.0010  0.5585  0.0000 -0.0000  0.5000 -0.0000 -0.0000 -0.0000 -0.0160]
 [-0.3052  0.0006  0.1268  0.2887  0.0000  0.4082  0.3960  0.5441 -0.0000]
 [-0.3426  0.0006  0.3052  0.4082  0.0000 -0.2887 -0.2641 -0.3960  0.0000]
 [ 0.0010  0.5585  0.0000  0.0000 -0.5000  0.0000 -0.0000 -0.0000 -0.0160]
 [ 0.3052 -0.0006  0.1268 -0.2887 -0.0000 -0.4082 -0.3960  0.5441  0.0000]
 [-0.3426  0.0006 -0.3052  0.4082  0.0000 -0.2887 -0.2641  0.3960  0.0000]
 [-0.3050  0.1274  0.0000 -0.2887 -0.0000 -0.4082  0.3961 -0.0000  0.5441]
 [ 0.0000 -0.0000  0.5585  0.0000 -0.5000  0.0000 -0.0000 -0.0160  0.0000]
 [-0.3421  0.3058  0.0000 -0.4082 -0.0000  0.2887 -0.2641  0.0000 -0.3960]
 [ 0.3055  0.1263 -0.0000  0.2887  0.0000  0.4082 -0.3960  0.0000  0.5441]
 [ 0.0000 -0.0000  0.5585 -0.0000  0.5000  0.0000 -0.0000 -0.0160  0.0000]
 [-0.3432 -0.3046  0.0000 -0.4082 -0.0000  0.2887 -0.2640  0.0000  0.3961]]
	dif: 
 [[-0.0007  0.0000 -0.0000  0.0000  0.0000 -0.0000 -0.0000 -0.0000 -0.0000]
 [-0.0000  0.0000  0.0000  0.0000 -0.0000 -0.0000 -0.0000 -0.0000  0.0000]
 [-0.0000 -0.0007 -0.0000 -0.0000 -0.0000 -0.0000  0.0000 -0.0000 -0.0000]
 [ 0.0010 -0.0000  0.0000 -0.0000 -0.0000 -0.0000 -0.0000 -0.0000  0.0000]
 [ 0.0000  0.0006  0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000 -0.0000]
 [ 0.0000  0.0006  0.0000  0.0000  0.0000 -0.0000 -0.0000  0.0000  0.0000]
 [ 0.0010 -0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000 -0.0000  0.0000]
 [ 0.0000 -0.0006 -0.0000 -0.0000 -0.0000  0.0000  0.0000  0.0000  0.0000]
 [-0.0000  0.0006  0.0000  0.0000  0.0000 -0.0000  0.0000  0.0000  0.0000]
 [ 0.0002  0.0006  0.0000  0.0000 -0.0000  0.0000  0.0001 -0.0000 -0.0000]
 [ 0.0000 -0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000  0.0000  0.0000]
 [ 0.0006  0.0006  0.0000  0.0000 -0.0000  0.0000 -0.0000  0.0000  0.0000]
 [ 0.0002 -0.0006 -0.0000 -0.0000  0.0000 -0.0000  0.0001  0.0000  0.0000]
 [ 0.0000 -0.0000  0.0000 -0.0000 -0.0000  0.0000 -0.0000  0.0000  0.0000]
 [-0.0006  0.0006  0.0000  0.0000 -0.0000  0.0000  0.0000  0.0000  0.0000]]
	test_hessian_vs_cfour[methane-H_analytic].........................PASSED
```

**Comment by susilehtola** [comment]
I don't see how that is possible - that test failed even when I tried it with the ```CORE``` guess above. The SCF must have converged since the job isn't crashing, and CH4 is a simple closed-shell molecule so there shouldn't be any problems with different HF solutions. This is also confirmed by the finite difference tests that pass for that molecule.

Whatever this test job is testing, its problems appear to be unrelated to this merge request. IMHO it shouldn't even be checking the vibrational frequencies and eigenmodes, but rather the raw Hessian which is way less ambiguous. From the output it is not clear what it is actually checking. Looks like they are eigenvectors, but the code is massaging them somehow, like you said. As you can see, the differences are arising only from the degenerate modes, so the related test code must not be working properly. The problem is that comparing eigenvectors for degenerate modes, you have to allow not only for rotations among the vectors, but also flipping the sign of the individual eigenvectors. Maybe this is not taken into account in the test code and that's what is causing the problems?

**Comment by loriab** [comment]
` 'SAD_SPIN_AVERAGE': True,` heals the test

**Comment by susilehtola** [comment]
> ` 'SAD_SPIN_AVERAGE': True,` heals the test

It's already ```True``` by default, and I didn't see anything in the pytest code that would change it. And the test similarly failed above with the ```CORE``` guess.

**Comment by JonathonMisiewicz** [comment]
Have you already checked for the possibility of non-deterministic pass/fail? That'd explain the "SAD_SPIN_AVERAGE" results.

**Comment by susilehtola** [comment]
@JonathonMisiewicz I'm not sure what you mean, but that's exactly what I've been arguing for above :D The problem is that since the failing test is a python one, I'm unable to run it locally.

**Comment by loriab** [comment]
can't run the pytest tests locally?
* `cd <objdir>`
* adjust PATH for psiapi http://psicode.org/psi4manual/master/build_planning.html#how-to-configure-paths-for-psiapi via `<objdir>/bin/psi --psiapi-path`
* get pytest via `conda install pytest` or `pip install pytest`
* `pytest ../tests/pytest -k methane-H_analytic`
the pytest tests can also be run via `psi4.test()`, `psi4 --test`, or `make pytest` but the above has more control

Yes, confirmed non-deterministic, so above was a red herring, sorry. But something has made it much less stable than before.

**Comment by susilehtola** [comment]
Cool. I've opened up a tracker so that the problems in the test are documented.

I see no reason not to merge.

File: psi4/src/psi4/libscf_solver/sad.cc [hunk 2]
```diff
@@ -247,14 +252,85 @@ SharedMatrix SADGuess::form_D_AO() {
...
+            }
+        } else {
+            // Conventional occupations
+            nocc_a = std::round(nalpha[index]);
+            occ_a = std::make_shared<Vector>("Alpha occupation", nocc_a);
+            for (size_t x = 0; x < nocc_a; x++) occ_a->set(x, 1.0);
+
+            nocc_b = std::round(nbeta[index]);
+            occ_b = std::make_shared<Vector>("Beta occupation", nocc_b);
+            for (size_t x = 0; x < nocc_b; x++) occ_b->set(x, 1.0);
```
Comment (dgasmith) [inline psi4/src/psi4/libscf_solver/sad.cc]: Is the history incorrect here, this should be merged with another PR no?

File: tests/scf-coverage/input.dat [hunk 3]
```diff
@@ -11,7 +11,7 @@ set d_convergence 8
 set stability_analysis check
 set basis cc-pVDZ
 set scf_type out_of_core
-set damping_percentage 50
+set damping_percentage 0
```
Comment (dgasmith) [inline tests/scf-coverage/input.dat]: Can you simply remove this option?

File: psi4/src/psi4/libscf_solver/sad.cc (hunk 2 above)
Comment (susilehtola) [reply psi4/src/psi4/libscf_solver/sad.cc]: It **is** correct - if you look down in the file, you'll see that this code has just been refactored. It used to be inside the atomic solver, but it's simpler to determine the occupations along with the spin state in the driver and just pass the occupation arrays to the solver since it's all it needs anyway.

File: tests/scf-coverage/input.dat (hunk 3 above)
Comment (susilehtola) [reply tests/scf-coverage/input.dat]: Yeah sure

File: psi4/src/psi4/libscf_solver/sad.cc [hunk 4]
```diff
@@ -167,35 +167,42 @@ SharedMatrix SADGuess::form_D_AO() {
...
     // For 37 to 86, save for f-block: Atomic, Molecular, & Optical Physics Handbook, Ed. Gordon W. F. Drake, American
     // Institute of Physics, Woodbury, New York, USA, 1996.
-    const int reference_S[] = {0, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 3, 6, 5, 4, 3, 2,
-                               1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 5, 6, 5, 4, 3, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1,
-                               0, 3, 4, 5, 6, 7, 8, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0};
-    const int MAX_Z = 86;
+    const std::vector<int> reference_S = {0, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1,
+                                          2, 3, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 5, 6, 5,
+                                          4, 3, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 0, 3, 4, 5, 6, 7, 8, 5,
+                                          4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0};
```
Comment (andysim) [inline psi4/src/psi4/libscf_solver/sad.cc]: Would you mind reformatting as
```
// clang-format off
const std::vector<int> reference_S = { 0,
1,                                                                                           0,
1, 0,                                                                         1, 2, 3, 2, 1, 0,
1, 0,                                                                         1, 2, 3, 2, 1, 0,
1, 0,                                           1, 2, 3, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0,
1, 0,                                           1, 2, 5, 6, 5, 4, 3, 0, 1, 0, 1, 2, 3, 2, 1, 0, 
1, 0, 1, 0, 3, 4, 5, 6, 7, 8, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0 };
// clang-format on
```
for readability, please?

File: psi4/src/psi4/libscf_solver/sad.cc [hunk 5]
```diff
@@ -167,35 +167,42 @@ SharedMatrix SADGuess::form_D_AO() {
...
-                            nhigh[A], nalpha[A], nbeta[A]);
+
+        if (options_.get_bool("SAD_SPIN_AVERAGE")) {
+            // Spin-averaged occupations
+            nalpha[A] = nbeta[A] = 0.5 * nelec[A];
+        } else {
+            // Target ground spin state
+            if (Z >= reference_S.size()) {
+                std::ostringstream err;
+                err << " Only atoms up to Z = " << reference_S.size() - 1
```
Comment (andysim) [inline psi4/src/psi4/libscf_solver/sad.cc]: I don't think the `-1` is needed here, because of the extra element at the front

File: psi4/src/psi4/libscf_solver/sad.cc (hunk 4 above)
Comment (susilehtola) [reply psi4/src/psi4/libscf_solver/sad.cc]: Done, thanks!

File: psi4/src/psi4/libscf_solver/sad.cc (hunk 5 above)
Comment (susilehtola) [reply psi4/src/psi4/libscf_solver/sad.cc]: It is. The first element is padding for ```Z=0``` which doesn't exist. The spin state for ```Z``` is ```reference_S[Z]```. The last valid ```Z``` is the one which gives ```reference_S[reference_S.size()-1]``` i.e. ```Z=reference_S.size()-1```.

**Comment by susilehtola** [comment]
Looks like the last Travis build failed due to an issue with setting up the build root. Any way to re-trigger the build?

**Comment by susilehtola** [comment]
Sorry for the feature creep, but due to the reorganization and the new default, I thought it'd be best to fix #1473 here as well.

I think the last patch should make the C++ side correct also for ECPs, but it appears that the atomic basis sets passed to the code still lack the ECP information. Does anyone know why?

**Comment by susilehtola** [comment]
ECP calculations work as well as ordinary ones, so this can finally be merged.

Here's the HF/def2-SVP calculation on Xe
```
  Initial Atomic UHF Energy:    -315.3634326754

                                         Total Energy            Delta E              Density RMS

  @Atomic UHF iteration   1 energy:  -280.36015363156582       35.00327904384721     0.17438162869310
  @Atomic UHF iteration   2 energy:  -260.70364549276712       19.65650813879870     0.20108587638483
  @Atomic UHF iteration   3 energy:  -320.59517730813167      -59.89153181536454     0.01619483472434
  @Atomic UHF iteration   4 energy:  -320.87131145782928       -0.27613414969761     0.00341002372407
  @Atomic UHF iteration   5 energy:  -320.87922264084102       -0.00791118301174     0.00056898557067
  @Atomic UHF iteration   6 energy:  -320.87957561560631       -0.00035297476529     0.00008011419372
  @Atomic UHF iteration   7 energy:  -320.87958210143461       -0.00000648582829     0.00000325221580
  @Atomic UHF Final Energy for atom XE:  -320.87958210143461
Finished UHF Computation!

  ==> Iterations <==

                           Total Energy        Delta E     RMS |[F,P]|

   @DF-RHF iter   0:  -324.53314738271274   -3.24533e+02   0.00000e+00 
   @DF-RHF iter   1:  -328.19660724674617   -3.66346e+00   2.76741e-02 DIIS
   @DF-RHF iter   2:  -328.28542700545580   -8.88198e-02   1.26940e-02 DIIS
   @DF-RHF iter   3:  -328.29832629846925   -1.28993e-02   5.50180e-04 DIIS
   @DF-RHF iter   4:  -328.29835689704061   -3.05986e-05   4.43928e-05 DIIS
   @DF-RHF iter   5:  -328.29835733022713   -4.33187e-07   3.97861e-06 DIIS
   @DF-RHF iter   6:  -328.29835733491979   -4.69265e-09   2.37557e-07 DIIS
  Energy converged.
```
Works as well as in UGBS
```
  Initial Atomic UHF Energy:    -5427.3786402080

                                         Total Energy            Delta E              Density RMS

  @Atomic UHF iteration   1 energy: -6483.96999842770038    -1056.59135821971540     1.04255411296391
  @Atomic UHF iteration   2 energy: -6453.45541954978944       30.51457887791094     0.40784617759788
  @Atomic UHF iteration   3 energy: -7117.90198205644811     -664.44656250665867     0.36315248442566
  @Atomic UHF iteration   4 energy: -7267.18887860081304     -149.28689654436494     0.12758818454125
  @Atomic UHF iteration   5 energy: -7299.46498016141504      -32.27610156060200     0.01509870939289
  @Atomic UHF iteration   6 energy: -7299.96418197356979       -0.49920181215475     0.00107318606111
  @Atomic UHF iteration   7 energy: -7299.97095115772936       -0.00676918415957     0.00050931817172
  @Atomic UHF iteration   8 energy: -7299.97187492498688       -0.00092376725752     0.00008205774167
  @Atomic UHF iteration   9 energy: -7299.97190856036286       -0.00003363537598     0.00000350327835
  @Atomic UHF iteration  10 energy: -7299.97190876705281       -0.00000020668995     0.00000049737307
  @Atomic UHF Final Energy for atom XE: -7299.97190876705281
Finished UHF Computation!

  ==> Iterations <==

                           Total Energy        Delta E     RMS |[F,P]|

   @DF-RHF iter   0: -7255.00101065488798   -7.25500e+03   0.00000e+00 
   @DF-RHF iter   1: -7259.15010997926129   -4.14910e+00   1.35868e-02 DIIS
   @DF-RHF iter   2: -7259.17972994142656   -2.96200e-02   3.74973e-03 DIIS
   @DF-RHF iter   3: -7259.18676197489549   -7.03203e-03   1.53495e-03 DIIS
   @DF-RHF iter   4: -7259.18756584481798   -8.03870e-04   4.30106e-05 DIIS
   @DF-RHF iter   5: -7259.18756735798888   -1.51317e-06   2.08951e-06 DIIS
   @DF-RHF iter   6: -7259.18756736888918   -1.09003e-08   3.49470e-07 DIIS
  Energy converged.
```

**Comment by dgasmith** [comment]
Cool, I think this looks great! Thank you for patching SAD up.

@psi4/maintainers @CDSherrill  Any last comments here before merging?

**Comment by loriab** [comment]
@susilehtola if you get a chance, could you try running @PeterKraus's new sapt-ecp test from #1350? I think that and this PR may have conflicted in that I get `Fatal Error: Atom has more electrons than basis functions.` for the *non-ECP* case. When I switch to `set guess core` (and back off decimals to 6), it runs and passes.

**Comment by susilehtola** [comment]
@loriab ah, it's a minimal basis issue; I already ran into the same issue in ERKALE but forgot to put the patch in Psi4.

**Comment by susilehtola** [comment]
Fix in #1481. The sapt-ecp runs but gives the wrong result regardless of the guess, see #1350.
